File: Sources/AppCore/Views/AppSettingsPanel.swift
import SwiftUI

/// The system-wide App Settings panel. Mirrors `CounterSettingsPanel`'s
/// floating-overlay chrome (anchored under the header, drawn on top, `DONE` to
/// close) and reuses `SettingsField` + `CounterTheme` tokens so it feels native.
///
/// Bound directly to the shared `AppSettings` store, so each control writes
/// straight through to the persisted default. Holds no local edit state: unlike
/// the per-counter panel there is no save/cancel — toggling is immediate.
public struct AppSettingsPanel: View {
    @ObservedObject var settings: AppSettings
    /// Room available below the anchoring header, supplied by `HeaderAnchoredOverlay`.
    /// The field body scrolls within this bound so the pinned DONE stays reachable.
    let availableHeight: CGFloat
    let onOpenList: () -> Void
    let onClose: () -> Void
    /// Whether the SOUND & HAPTICS section starts expanded. App Settings has
    /// no per-counter override state to key off, so it collapses by default in
    /// production; feedback-focused isolated scenarios opt specific cases open.
    @State private var showFeedback: Bool

    public init(settings: AppSettings,
                availableHeight: CGFloat,
                initiallyExpandedFeedback: Bool = false,
                onOpenList: @escaping () -> Void,
                onClose: @escaping () -> Void) {
        self.settings = settings
        self.availableHeight = availableHeight
        self.onOpenList = onOpenList
        self.onClose = onClose
        _showFeedback = State(initialValue: initiallyExpandedFeedback)
    }

    public var body: some View {
        // Cap the card to the room below the anchor, less the top inset, the
        // card's own vertical padding, and a bottom breathing margin.
        let maxCardHeight = max(160, availableHeight - 12 - 40 - 24)

        return VStack(alignment: .leading, spacing: 18) {
            header

            BoundedScroll {
                VStack(alignment: .leading, spacing: 18) {
                    SettingsField("HANDEDNESS") {
                        handednessControl
                    }

                    FeedbackDisclosureToggle(expanded: $showFeedback,
                                             title: "SOUND & HAPTICS",
                                             identifier: "app-settings-feedback-toggle")

                    if showFeedback {
                        SettingsField("SOUND ON CHANGE") {
                            optionPicker(options: SoundOption.allCases,
                                         selected: settings.soundOption,
                                         label: { $0.label },
                                         onSelect: { settings.soundOption = $0 },
                                         id: "app-settings-sound")
                        }

                        SettingsField("INCREMENT HAPTIC") {
                            optionPicker(options: HapticOption.allCases,
                                         selected: settings.incrementHapticOption,
                                         label: { $0.label },
                                         onSelect: { settings.incrementHapticOption = $0 },
                                         id: "app-settings-increment-haptic")
                        }

                        SettingsField("DECREMENT HAPTIC") {
                            optionPicker(options: HapticOption.allCases,
                                         selected: settings.decrementHapticOption,
                                         label: { $0.label },
                                         onSelect: { settings.decrementHapticOption = $0 },
                                         id: "app-settings-decrement-haptic")
                        }
                    }

                    allCountersButton
                }
            }
        }
        .frame(maxHeight: maxCardHeight, alignment: .top)
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .top)
        .background(CounterTheme.panel)
        .overlay(Rectangle().stroke(CounterTheme.lineStrong, lineWidth: 1))
        .padding(.horizontal, 22)
        .padding(.top, 12)
    }

    // A horizontally-scrolling row of selectable pills — one per option — with
    // the active one filled in the accent color. Reused for both the sound and
    // haptic choices; scrolls when the options overflow the panel width.
    private func optionPicker<T: Hashable>(options: [T],
                                           selected: T,
                                           label: @escaping (T) -> String,
                                           onSelect: @escaping (T) -> Void,
                                           id: String) -> some View {
        ScrollViewReader { proxy in
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(options, id: \.self) { option in
                        let isSelected = option == selected
                        Button(action: { onSelect(option) }) {
                            Text(label(option))
                                .font(.system(size: 12, weight: .heavy, design: .monospaced))
                                .tracking(1)
                                .foregroundColor(isSelected ? CounterTheme.onAccent : CounterTheme.ink)
                                .padding(.horizontal, 14)
                                .padding(.vertical, 9)
                                .background(isSelected ? CounterTheme.accent : Color.clear)
                                .overlay(Rectangle().stroke(isSelected ? CounterTheme.accent : CounterTheme.line, lineWidth: 1))
                        }
                        .buttonStyle(.plain)
                        .id(option)
                        .accessibilityIdentifier("\(id)-\(label(option).lowercased())")
                    }
                }
                .padding(.vertical, 2)
            }
            // Keep the active option visible even when it sits past the fold —
            // otherwise a selection like the last sound reads as "nothing chosen".
            .onAppear { proxy.scrollTo(selected, anchor: .center) }
        }
        .accessibilityIdentifier(id)
    }

    // A two-way segmented Left / Right control writing `defaultLeftHanded`.
    private var handednessControl: some View {
        HStack(spacing: 0) {
            handednessOption(label: "LEFT", isLeft: true)
            handednessOption(label: "RIGHT", isLeft: false)
        }
        .overlay(Rectangle().stroke(CounterTheme.line, lineWidth: 1))
        .accessibilityIdentifier("app-settings-handedness")
    }

    private func handednessOption(label: String, isLeft: Bool) -> some View {
        let selected = settings.defaultLeftHanded == isLeft
        return Button(action: { settings.defaultLeftHanded = isLeft }) {
            Text(label)
                .font(.system(size: 12, weight: .heavy, design: .monospaced))
                .tracking(1)
                .foregroundColor(selected ? CounterTheme.onAccent : CounterTheme.ink)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 10)
                .background(selected ? CounterTheme.accent : Color.clear)
        }
        .buttonStyle(.plain)
    }

    private var allCountersButton: some View {
        Button(action: onOpenList) {
            HStack {
                Text("ALL COUNTERS")
                    .font(.system(size: 13, weight: .heavy, design: .monospaced))
                    .tracking(1)
                    .foregroundColor(CounterTheme.ink)
                Spacer()
                Image(systemName: "chevron.right")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundColor(CounterTheme.inkMuted)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 12)
            .padding(.horizontal, 12)
            .overlay(Rectangle().stroke(CounterTheme.lineStrong, lineWidth: 1))
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("app-settings-list")
    }

    private var header: some View {
        HStack {
            Text("APP SETTINGS")
                .font(.system(size: 12, weight: .bold, design: .monospaced))
                .tracking(1.4)
                .foregroundColor(CounterTheme.inkMuted)
            Spacer()
            Button(action: onClose) {
                Text("DONE")
                    .font(.system(size: 12, weight: .heavy, design: .monospaced))
                    .tracking(1)
                    .foregroundColor(CounterTheme.onAccent)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 8)
                    .background(CounterTheme.accent)
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("app-settings-close")
        }
    }
}

File: android/app/src/main/java/com/codeyam/android/ui/AppSettingsPanel.kt
package com.codeyam.android.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.codeyam.android.BuildConfig
import com.codeyam.android.model.HapticOption
import com.codeyam.android.model.SoundOption

/**
 * The system-wide App Settings panel. Mirrors the per-counter panel's chrome
 * (anchored under the header, DONE to close) but holds no local edit state:
 * every control writes straight through to the persisted default, so there is no
 * save/cancel — toggling is immediate. Ports iOS `AppSettingsPanel`.
 *
 * A pure value-driven view: it takes the four current settings values plus four
 * change callbacks and renders exactly what it is handed. It does NOT read
 * `AppSettings` (a framework-free model object with nothing observable to
 * subscribe to), so the CALLER must supply values that recompose when they
 * change — see `CounterScreenState.defaultLeftHanded` and its sibling
 * revision-keyed getters. Reaching into the store from here instead would
 * reintroduce the stale-highlight gap this shape exists to close.
 */
@Composable
fun AppSettingsPanel(
    leftHanded: Boolean,
    soundOption: SoundOption,
    incrementHapticOption: HapticOption,
    decrementHapticOption: HapticOption,
    onLeftHandedChange: (Boolean) -> Unit,
    onSoundChange: (SoundOption) -> Unit,
    onIncrementHapticChange: (HapticOption) -> Unit,
    onDecrementHapticChange: (HapticOption) -> Unit,
    availableHeight: Dp,
    onOpenList: () -> Unit,
    onClose: () -> Unit,
    modifier: Modifier = Modifier,
) {
    SettingsPanelCard(
        title = "APP SETTINGS",
        doneIdentifier = "app-settings-close",
        onDone = onClose,
        availableHeight = availableHeight,
        modifier = modifier,
    ) {
        SettingsField("HANDEDNESS") {
            HandednessControl(
                leftHanded = leftHanded,
                onSelect = onLeftHandedChange,
            )
        }

        // App Settings has no per-counter override state to justify a collapsed
        // resting state, so the sound/haptic rows are always visible rather than
        // hidden behind a disclosure.
        SettingsField("SOUND ON CHANGE") {
            OptionPicker(
                options = SoundOption.entries,
                selected = soundOption,
                onSelect = onSoundChange,
                idPrefix = "app-settings-sound",
                optionLabel = { it.label },
            )
        }

        SettingsField("INCREMENT HAPTIC") {
            OptionPicker(
                options = HapticOption.entries,
                selected = incrementHapticOption,
                onSelect = onIncrementHapticChange,
                idPrefix = "app-settings-increment-haptic",
                optionLabel = { it.label },
            )
        }

        SettingsField("DECREMENT HAPTIC") {
            OptionPicker(
                options = HapticOption.entries,
                selected = decrementHapticOption,
                onSelect = onDecrementHapticChange,
                idPrefix = "app-settings-decrement-haptic",
                optionLabel = { it.label },
            )
        }

        AllCountersButton(onClick = onOpenList)

        // Build identity, so "which version am I on?" is a glance rather than an
        // adb command. versionName is always 1.0; the versionCode is what tells
        // internal-test builds apart.
        Text(
            text = "v${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
            fontSize = 10.sp,
            fontFamily = FontFamily.Monospace,
            letterSpacing = 0.5.sp,
            color = CounterColors.inkMuted,
            textAlign = TextAlign.Center,
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 10.dp)
                .semantics { contentDescription = "app-settings-version" },
        )
    }
}

/** A two-way segmented LEFT / RIGHT control writing `defaultLeftHanded`. */
@Composable
private fun HandednessControl(leftHanded: Boolean, onSelect: (Boolean) -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .border(1.dp, CounterColors.line, RectangleShape)
            .semantics { contentDescription = "app-settings-handedness" },
    ) {
        HandednessOption(
            label = "LEFT",
            identifier = "app-settings-handedness-left",
            isSelected = leftHanded,
            onClick = { onSelect(true) },
        )
        HandednessOption(
            label = "RIGHT",
            identifier = "app-settings-handedness-right",
            isSelected = !leftHanded,
            onClick = { onSelect(false) },
        )
    }
}

@Composable
private fun androidx.compose.foundation.layout.RowScope.HandednessOption(
    label: String,
    identifier: String,
    isSelected: Boolean,
    onClick: () -> Unit,
) {
    val interaction = remember { MutableInteractionSource() }
    Text(
        text = label,
        fontSize = 12.sp,
        fontWeight = FontWeight.Black,
        fontFamily = FontFamily.Monospace,
        letterSpacing = 1.sp,
        color = if (isSelected) CounterColors.onAccent else CounterColors.ink,
        textAlign = TextAlign.Center,
        modifier = Modifier
            .weight(1f)
            .background(if (isSelected) CounterColors.accent else Color.Transparent)
            .clickable(interactionSource = interaction, indication = null, onClick = onClick)
            .padding(vertical = 10.dp)
            // Per-option identifier + selection state so a tap can target LEFT /
            // RIGHT individually and tests (and TalkBack) can tell which is active.
            .semantics {
                contentDescription = identifier
                selected = isSelected
            },
    )
}

/** The row that opens the all-counters list overlay. */
@Composable
private fun AllCountersButton(onClick: () -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .border(1.dp, CounterColors.lineStrong, RectangleShape)
            .clickable(interactionSource = interaction, indication = null, onClick = onClick)
            .padding(vertical = 12.dp, horizontal = 12.dp)
            .semantics { contentDescription = "app-settings-list" },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            text = "ALL COUNTERS",
            fontSize = 13.sp,
            fontWeight = FontWeight.Black,
            fontFamily = FontFamily.Monospace,
            letterSpacing = 1.sp,
            color = CounterColors.ink,
        )
        Text(
            text = "›",
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
            color = CounterColors.inkMuted,
        )
    }
}
